File: app/src/main/java/com/boardgamegeek/service/SyncPlaysUpload.java
package com.boardgamegeek.service;

import android.accounts.Account;
import android.app.PendingIntent;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.content.SyncResult;
import android.database.Cursor;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.NotificationCompat.Action;
import android.util.Pair;

import com.boardgamegeek.R;
import com.boardgamegeek.auth.Authenticator;
import com.boardgamegeek.io.BggService;
import com.boardgamegeek.model.Play;
import com.boardgamegeek.model.PlayDeleteResponse;
import com.boardgamegeek.model.PlaySaveResponse;
import com.boardgamegeek.model.Player;
import com.boardgamegeek.model.builder.PlayBuilder;
import com.boardgamegeek.model.persister.PlayPersister;
import com.boardgamegeek.provider.BggContract;
import com.boardgamegeek.provider.BggContract.Collection;
import com.boardgamegeek.provider.BggContract.Games;
import com.boardgamegeek.provider.BggContract.Plays;
import com.boardgamegeek.ui.PlaysActivity;
import com.boardgamegeek.util.ActivityUtils;
import com.boardgamegeek.util.CursorUtils;
import com.boardgamegeek.util.HttpUtils;
import com.boardgamegeek.util.NotificationUtils;
import com.boardgamegeek.util.PresentationUtils;
import com.boardgamegeek.util.SelectionBuilder;
import com.boardgamegeek.util.StringUtils;

import java.util.List;

import hugo.weaving.DebugLog;
import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Request.Builder;

public class SyncPlaysUpload extends SyncUploadTask {
	public static final String GEEK_PLAY_URL = "https://www.boardgamegeek.com/geekplay.php";
	private OkHttpClient httpClient;
	private PlayPersister persister;
	private long currentInternalIdForMessage;
	private int currentGameIdForMessage;
	private String currentGameNameForMessage;

	@DebugLog
	public SyncPlaysUpload(Context context, BggService service) {
		super(context, service);
	}

	@DebugLog
	@Override
	public int getSyncType() {
		return SyncService.FLAG_SYNC_PLAYS_UPLOAD;
	}

	@DebugLog
	@Override
	protected int getNotificationTitleResId() {
		return R.string.sync_notification_title_play_upload;
	}

	@DebugLog
	@Override
	protected Intent getNotificationIntent() {
		return new Intent(context, PlaysActivity.class);
	}

	@DebugLog
	@Override
	protected String getNotificationMessageTag() {
		return NotificationUtils.TAG_UPLOAD_PLAY;
	}

	@DebugLog
	@Override
	protected String getNotificationErrorTag() {
		return NotificationUtils.TAG_UPLOAD_PLAY_ERROR;
	}

	@DebugLog
	@Override
	public void execute(Account account, @NonNull SyncResult syncResult) {
		httpClient = HttpUtils.getHttpClientWithAuth(context);
		persister = new PlayPersister(context);

		deletePendingPlays(syncResult);
		updatePendingPlays(syncResult);
		if (SyncService.isPlaysSyncUpToDate(context)) {
			SyncService.calculateAndUpdateHIndex(context);
		}
	}

	@DebugLog
	@Override
	public int getNotificationSummaryMessageId() {
		return R.string.sync_notification_plays_upload;
	}

	@DebugLog
	private void updatePendingPlays(@NonNull SyncResult syncResult) {
		Cursor cursor = null;
		try {
			cursor = context.getContentResolver().query(Plays.CONTENT_SIMPLE_URI,
				PlayBuilder.PLAY_PROJECTION_WITH_ID,
				Plays.UPDATE_TIMESTAMP + ">0",
				null,
				Plays.UPDATE_TIMESTAMP);
			int playCount = cursor != null ? cursor.getCount() : 0;
			updateProgressNotificationAsPlural(R.plurals.sync_notification_progress_update, playCount, playCount);

			while (cursor != null && cursor.moveToNext()) {
				if (isCancelled()) break;
				if (wasSleepInterrupted(1000)) break;

				currentInternalIdForMessage = CursorUtils.getLong(cursor, Plays._ID, BggContract.INVALID_ID);
				Play play = PlayBuilder.fromCursor(cursor);
				Cursor playerCursor = PlayBuilder.queryPlayers(context, currentInternalIdForMessage);
				try {
					PlayBuilder.addPlayers(playerCursor, play);
				} finally {
					if (playerCursor != null) playerCursor.close();
				}

				PlaySaveResponse response = postPlayUpdate(play);
				if (response == null) {
					syncResult.stats.numIoExceptions++;
					notifyUploadError(context.getString(R.string.msg_play_update_null_response));
				} else if (response.hasAuthError()) {
					syncResult.stats.numAuthExceptions++;
					Authenticator.clearPassword(context);
					break;
				} else if (response.hasInvalidIdError()) {
					notifyUploadError(PresentationUtils.getText(context, R.string.msg_play_update_bad_id, play.playId));
				} else if (response.hasError()) {
					notifyUploadError(response.getErrorMessage());
				} else if (response.getPlayCount() <= 0) {
					syncResult.stats.numIoExceptions++;
					notifyUploadError(context.getString(R.string.msg_play_update_null_response));
				} else {
					CharSequence message = play.playId > 0 ?
						PresentationUtils.getText(context, R.string.msg_play_updated) :
						PresentationUtils.getText(context, R.string.msg_play_added, getPlayCountDescription(response.getPlayCount(), play.quantity));

					play.playId = response.getPlayId();
					play.dirtyTimestamp = 0;
					play.updateTimestamp = 0;
					play.deleteTimestamp = 0;
					currentGameIdForMessage = play.gameId;
					currentGameNameForMessage = play.gameName;

					notifyUser(play, message);
					persister.save(play, currentInternalIdForMessage, false);

					updateGamePlayCount(play);
				}
			}
		} finally {
			if (cursor != null && !cursor.isClosed()) {
				cursor.close();
			}
		}
	}

	private void notifyUser(Play play, CharSequence message) {
		Pair<String, String> imageUrls = queryGameImageUrls(play);
		notifyUser(play.gameName, message, NotificationUtils.getIntegerId(currentInternalIdForMessage), imageUrls.first, imageUrls.second);
	}

	private Pair<String, String> queryGameImageUrls(Play play) {
		Pair<String, String> imageUrls = Pair.create("", "");
		Cursor gameCursor = context.getContentResolver().query(Games.buildGameUri(play.gameId),
			new String[] { Games.IMAGE_URL, Games.THUMBNAIL_URL }, null, null, null);
		try {
			if (gameCursor != null && gameCursor.moveToFirst()) {
				imageUrls = Pair.create(gameCursor.getString(0), gameCursor.getString(1));
			}
		} finally {
			if (gameCursor != null) gameCursor.close();
		}
		return imageUrls;
	}

	@DebugLog
	private void deletePendingPlays(@NonNull SyncResult syncResult) {
		Cursor cursor = null;
		try {
			cursor = context.getContentResolver().query(Plays.CONTENT_SIMPLE_URI,
				PlayBuilder.PLAY_PROJECTION_WITH_ID,
				Plays.DELETE_TIMESTAMP + ">0",
				null,
				Plays.DELETE_TIMESTAMP);
			int playCount = cursor != null ? cursor.getCount() : 0;
			updateProgressNotificationAsPlural(R.plurals.sync_notification_progress_delete, playCount, playCount);

			while (cursor != null && cursor.moveToNext()) {
				if (isCancelled()) break;
				currentInternalIdForMessage = CursorUtils.getLong(cursor, Plays._ID, BggContract.INVALID_ID);
				Play play = PlayBuilder.fromCursor(cursor);
				if (play.playId > 0) {
					if (wasSleepInterrupted(1000)) break;
					PlayDeleteResponse response = postPlayDelete(play.playId);
					if (response == null) {
						syncResult.stats.numIoExceptions++;
						notifyUploadError(context.getString(R.string.msg_play_update_null_response));
					} else if (response.isSuccessful()) {
						deletePlay(currentInternalIdForMessage);
						updateGamePlayCount(play);
						notifyUserOfDelete(R.string.msg_play_deleted, play);
					} else if (response.hasInvalidIdError()) {
						deletePlay(currentInternalIdForMessage);
						notifyUserOfDelete(R.string.msg_play_deleted, play);
					} else if (response.hasAuthError()) {
						syncResult.stats.numAuthExceptions++;
						Authenticator.clearPassword(context);
					} else {
						syncResult.stats.numIoExceptions++;
						notifyUploadError(response.getErrorMessage());
					}
				} else {
					deletePlay(currentInternalIdForMessage);
					notifyUserOfDelete(R.string.msg_play_deleted_draft, play);
				}
			}
		} finally {
			if (cursor != null && !cursor.isClosed()) {
				cursor.close();
			}
		}
	}

	@DebugLog
	private void updateGamePlayCount(@NonNull Play play) {
		ContentResolver resolver = context.getContentResolver();
		Cursor cursor = null;
		try {
			cursor = resolver.query(Plays.CONTENT_SIMPLE_URI,
				new String[] { Plays.SUM_QUANTITY },
				String.format("%s=? AND %s", Plays.OBJECT_ID, SelectionBuilder.whereZeroOrNull(Plays.DELETE_TIMESTAMP)),
				new String[] { String.valueOf(play.gameId) },
				null);
			if (cursor != null && cursor.moveToFirst()) {
				int newPlayCount = cursor.getInt(0);
				ContentValues values = new ContentValues();
				values.put(Collection.NUM_PLAYS, newPlayCount);
				resolver.update(Games.buildGameUri(play.gameId), values, null, null);
			}
		} finally {
			if (cursor != null) {
				cursor.close();
			}
		}
	}

	@DebugLog
	@Nullable
	private PlaySaveResponse postPlayUpdate(@NonNull Play play) {
		FormBody.Builder builder = new FormBody.Builder()
			.add("ajax", "1")
			.add("action", "save")
			.add("version", "2")
			.add("objecttype", "thing");
		if (play.playId > 0) {
			builder.add("playid", String.valueOf(play.playId));
		}
		builder.add("objectid", String.valueOf(play.gameId))
			.add("playdate", play.getDate())
			.add("dateinput", play.getDate())
			.add("length", String.valueOf(play.length))
			.add("location", play.location)
			.add("quantity", String.valueOf(play.quantity))
			.add("incomplete", play.Incomplete() ? "1" : "0")
			.add("nowinstats", play.NoWinStats() ? "1" : "0")
			.add("comments", play.comments);
		List<Player> players = play.getPlayers();
		for (int i = 0; i < players.size(); i++) {
			Player player = players.get(i);
			builder
				.add(getMapKey(i, "playerid"), "player_" + i)
				.add(getMapKey(i, "name"), player.name)
				.add(getMapKey(i, "username"), player.username)
				.add(getMapKey(i, "color"), player.color)
				.add(getMapKey(i, "position"), player.startposition)
				.add(getMapKey(i, "score"), player.score)
				.add(getMapKey(i, "rating"), String.valueOf(player.rating))
				.add(getMapKey(i, "new"), String.valueOf(player.new_))
				.add(getMapKey(i, "win"), String.valueOf(player.win));
		}

		Request request = new Builder()
			.url(GEEK_PLAY_URL)
			.post(builder.build())
			.build();
		return new PlaySaveResponse(httpClient, request);
	}

	@DebugLog
	@NonNull
	private static String getMapKey(int index, String key) {
		return "players[" + index + "][" + key + "]";
	}

	@DebugLog
	@Nullable
	private PlayDeleteResponse postPlayDelete(int playId) {
		FormBody.Builder builder = new FormBody.Builder()
			.add("ajax", "1")
			.add("action", "delete")
			.add("playid", String.valueOf(playId))
			.add("finalize", "1");

		Request request = new Builder()
			.url(GEEK_PLAY_URL)
			.post(builder.build())
			.build();
		return new PlayDeleteResponse(httpClient, request);
	}

	/**
	 * Deletes the specified play from the content provider
	 */
	@DebugLog
	private void deletePlay(long internalId) {
		persister.delete(internalId);
	}

	@DebugLog
	private String getPlayCountDescription(int count, int quantity) {
		String countDescription;
		switch (quantity) {
			case 1:
				countDescription = StringUtils.getOrdinal(count);
				break;
			case 2:
				countDescription = StringUtils.getOrdinal(count - 1) + " & " + StringUtils.getOrdinal(count);
				break;
			default:
				countDescription = StringUtils.getOrdinal(count - quantity + 1) + " - " + StringUtils.getOrdinal(count);
				break;
		}
		return countDescription;
	}

	@DebugLog
	private void notifyUserOfDelete(@StringRes int messageId, Play play) {
		notifyUser(play, PresentationUtils.getText(context, messageId, play.gameName));
	}

	@DebugLog
	@Override
	protected Action createMessageAction() {
		if (currentInternalIdForMessage != BggContract.INVALID_ID) {
			Intent intent = ActivityUtils.createRematchIntent(context,
				currentInternalIdForMessage,
				currentGameIdForMessage,
				currentGameNameForMessage, null, null);
			PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
			NotificationCompat.Action.Builder builder = new NotificationCompat.Action.Builder(
				R.drawable.ic_replay_black_24dp,
				context.getString(R.string.rematch),
				pendingIntent);
			return builder.build();
		}
		return null;
	}
}

File: app/src/main/java/com/boardgamegeek/service/SyncPlaysUpload.kt
package com.boardgamegeek.service

import android.app.PendingIntent
import android.content.Intent
import android.content.SyncResult
import android.os.Build
import androidx.annotation.StringRes
import androidx.core.app.NotificationCompat.Action
import com.boardgamegeek.BggApplication
import com.boardgamegeek.R
import com.boardgamegeek.auth.Authenticator
import com.boardgamegeek.entities.PlayEntity
import com.boardgamegeek.extensions.*
import com.boardgamegeek.provider.BggContract.Companion.INVALID_ID
import com.boardgamegeek.provider.BggContract.Games
import com.boardgamegeek.repository.PlayRepository
import com.boardgamegeek.ui.GamePlaysActivity
import com.boardgamegeek.ui.LogPlayActivity
import com.boardgamegeek.ui.PlayActivity
import com.boardgamegeek.ui.PlaysActivity
import kotlinx.coroutines.runBlocking
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request.Builder
import kotlin.time.Duration.Companion.seconds

class SyncPlaysUpload(
    application: BggApplication,
    syncResult: SyncResult,
    private val repository: PlayRepository,
    private val httpClient: OkHttpClient,
) : SyncUploadTask(application, syncResult) {
    private var currentPlay = PlayForNotification()
    private val gameIds = mutableSetOf<Int>()

    inner class PlayForNotification(
        var internalId: Long = INVALID_ID.toLong(),
        val gameId: Int = INVALID_ID,
        val gameName: String = ""
    ) {
        var imageUrl: String = ""
        var thumbnailUrl: String = ""
        var heroImageUrl: String = ""
        var customPlayerSort: Boolean = false

        val internalIdAsInt = if (internalId < Int.MAX_VALUE) internalId.toInt() else (internalId % Int.MAX_VALUE).toInt()
    }

    override val syncType = SyncService.FLAG_SYNC_PLAYS_UPLOAD

    override val notificationTitleResId = R.string.sync_notification_title_play_upload

    override val summarySuffixResId = R.plurals.plays_suffix

    override val notificationSummaryIntent = context.intentFor<PlaysActivity>()

    override val notificationIntent: Intent
        get() = if (currentPlay.internalId == INVALID_ID.toLong())
            GamePlaysActivity.createIntent(
                context,
                currentPlay.gameId,
                currentPlay.gameName,
                currentPlay.imageUrl,
                currentPlay.thumbnailUrl,
                currentPlay.heroImageUrl
            )
        else
            PlayActivity.createIntent(context, currentPlay.internalId)

    override val notificationMessageTag = NotificationTags.UPLOAD_PLAY

    override val notificationErrorTag = NotificationTags.UPLOAD_PLAY_ERROR

    override val notificationSummaryMessageId = R.string.sync_notification_plays_upload

    override fun execute() {
        deletePendingPlays()
        updatePendingPlays()
        runBlocking {
            gameIds.forEach { repository.updateGamePlayCount(it) }
            repository.calculatePlayStats()
        }
    }

    private fun updatePendingPlays() {
        val pendingPlays = runBlocking { repository.getUpdatingPlays() }
        var currentNumberOfPlays = 0
        val totalNumberOfPlays = pendingPlays.size
        updateProgressNotificationAsPlural(R.plurals.sync_notification_plays_update, totalNumberOfPlays, totalNumberOfPlays)
        pendingPlays.forEach { play ->
            if (isCancelled) return
            if (wasSleepInterrupted(1.seconds, false)) return

            updateProgressNotificationAsPlural(
                R.plurals.sync_notification_plays_update_increment,
                totalNumberOfPlays,
                ++currentNumberOfPlays,
                totalNumberOfPlays,
            )

            try {
                val response = postPlayUpdate(play)
                when {
                    response.hasAuthError() -> {
                        syncResult.stats.numAuthExceptions++
                        Authenticator.clearPassword(context)
                        return
                    }
                    response.hasInvalidIdError() -> {
                        syncResult.stats.numConflictDetectedExceptions++
                        notifyUploadError(context.getText(R.string.msg_play_update_bad_id, play.playId))
                    }
                    response.hasError() -> {
                        syncResult.stats.numIoExceptions++
                        notifyUploadError(response.errorMessage)
                    }
                    response.playCount < 0 -> {
                        syncResult.stats.numIoExceptions++
                        notifyUploadError(context.getString(R.string.msg_play_update_null_response))
                    }
                    else -> {
                        syncResult.stats.numUpdates++
                        val message = when {
                            play.isSynced -> context.getText(R.string.msg_play_updated)
                            play.quantity > 0 -> context.getText(
                                R.string.msg_play_added_quantity,
                                getPlayCountDescription(response.playCount, play.quantity)
                            )
                            else -> context.getText(R.string.msg_play_added)
                        }

                        currentPlay = PlayForNotification(play.internalId, play.gameId, play.gameName)
                        notifyUser(play, message)

                        runBlocking { repository.markAsSynced(play.internalId, response.playId) }
                        gameIds += play.gameId
                    }
                }
            } catch (e: Exception) {
                syncResult.stats.numParseExceptions++
                notifyUploadError(e.localizedMessage.orEmpty())
            }
        }
    }

    private fun deletePendingPlays() {
        val deletedPlays = runBlocking { repository.getDeletingPlays() }
        var currentNumberOfPlays = 0
        val totalNumberOfPlays = deletedPlays.size
        updateProgressNotificationAsPlural(R.plurals.sync_notification_plays_delete, totalNumberOfPlays, totalNumberOfPlays)

        deletedPlays.forEach { play ->
            if (isCancelled) return
            if (wasSleepInterrupted(1.seconds, false)) return

            updateProgressNotificationAsPlural(
                R.plurals.sync_notification_plays_delete_increment,
                totalNumberOfPlays,
                ++currentNumberOfPlays,
                totalNumberOfPlays
            )

            try {
                currentPlay = PlayForNotification(play.internalId, play.gameId, play.gameName)
                if (play.isSynced) {
                    val response = postPlayDelete(play.playId)
                    when {
                        response.isSuccessful -> {
                            syncResult.stats.numDeletes++
                            runBlocking { repository.delete(play.internalId) }
                            gameIds += play.gameId
                            notifyUserOfDelete(R.string.msg_play_deleted, play)
                        }
                        response.hasInvalidIdError() -> {
                            syncResult.stats.numConflictDetectedExceptions++
                            runBlocking { repository.delete(play.internalId) }
                            notifyUserOfDelete(R.string.msg_play_deleted, play)
                        }
                        response.hasAuthError() -> {
                            syncResult.stats.numAuthExceptions++
                            Authenticator.clearPassword(context)
                            return
                        }
                        else -> {
                            syncResult.stats.numIoExceptions++
                            notifyUploadError(response.errorMessage)
                        }
                    }
                } else {
                    syncResult.stats.numDeletes++
                    runBlocking { repository.delete(play.internalId) }
                    gameIds += play.gameId
                    notifyUserOfDelete(R.string.msg_play_deleted_draft, play)
                }
            } catch (e: Exception) {
                syncResult.stats.numParseExceptions++
                notifyUploadError(e.localizedMessage.orEmpty())
            }
        }
    }

    @Suppress("SpellCheckingInspection")
    private fun postPlayUpdate(play: PlayEntity): PlaySaveResponse {
        val bodyBuilder = FormBody.Builder()
            .add("ajax", "1")
            .add("action", "save")
            .add("version", "2")
            .add("objecttype", "thing")
        if (play.playId > 0) {
            bodyBuilder.add("playid", play.playId.toString())
        }
        bodyBuilder.add("objectid", play.gameId.toString())
            .add("playdate", play.dateInMillis.asDateForApi())
            .add("dateinput", play.dateInMillis.asDateForApi())
            .add("length", play.length.toString())
            .add("location", play.location)
            .add("quantity", play.quantity.toString())
            .add("incomplete", if (play.incomplete) "1" else "0")
            .add("nowinstats", if (play.noWinStats) "1" else "0")
            .add("comments", play.comments)
        val players = play.players
        for (i in players.indices) {
            val player = players[i]
            bodyBuilder
                .add(getMapKey(i, "playerid"), "player_$i")
                .add(getMapKey(i, "name"), player.name)
                .add(getMapKey(i, "username"), player.username)
                .add(getMapKey(i, "color"), player.color)
                .add(getMapKey(i, "position"), player.startingPosition)
                .add(getMapKey(i, "score"), player.score)
                .add(getMapKey(i, "rating"), player.rating.toString())
                .add(getMapKey(i, "new"), if (player.isNew) "1" else "0")
                .add(getMapKey(i, "win"), if (player.isWin) "1" else "0")
        }

        val request = Builder()
            .url(GEEK_PLAY_URL)
            .post(bodyBuilder.build())
            .build()
        return PlaySaveResponse(httpClient, request)
    }

    private fun postPlayDelete(playId: Int): PlayDeleteResponse {
        @Suppress("SpellCheckingInspection")
        val bodyBuilder = FormBody.Builder()
            .add("ajax", "1")
            .add("action", "delete")
            .add("playid", playId.toString())
            .add("finalize", "1")

        val request = Builder()
            .url(GEEK_PLAY_URL)
            .post(bodyBuilder.build())
            .build()
        return PlayDeleteResponse(httpClient, request)
    }

    private fun getPlayCountDescription(count: Int, quantity: Int): String {
        return when (quantity) {
            1 -> count.toOrdinal()
            2 -> "${(count - 1).toOrdinal()} & ${count.toOrdinal()}"
            else -> "${(count - quantity + 1).toOrdinal()} - ${count.toOrdinal()}"
        }
    }

    private fun notifyUserOfDelete(@StringRes messageId: Int, play: PlayEntity) {
        context.cancelNotification(notificationMessageTag, currentPlay.internalIdAsInt.toLong())
        currentPlay.internalId = INVALID_ID.toLong()
        notifyUser(play, context.getText(messageId, play.gameName))
    }

    private fun notifyUser(play: PlayEntity, message: CharSequence) {
        if (play.gameId != INVALID_ID) {
            val gameCursor = context.contentResolver.query(
                Games.buildGameUri(play.gameId),
                arrayOf(
                    Games.Columns.IMAGE_URL,
                    Games.Columns.THUMBNAIL_URL,
                    Games.Columns.HERO_IMAGE_URL,
                    Games.Columns.CUSTOM_PLAYER_SORT,
                ),
                null,
                null,
                null
            )
            gameCursor?.use {
                if (it.moveToFirst()) {
                    currentPlay.imageUrl = it.getString(0).orEmpty()
                    currentPlay.thumbnailUrl = it.getString(1).orEmpty()
                    currentPlay.heroImageUrl = it.getString(2).orEmpty()
                    currentPlay.customPlayerSort = it.getInt(3) == 1
                }
            }
        }
        notifyUser(
            play.gameName,
            message,
            currentPlay.internalIdAsInt,
            currentPlay.heroImageUrl,
            currentPlay.thumbnailUrl,
            currentPlay.imageUrl,
        )
    }

    override fun createMessageAction(): Action? {
        if (currentPlay.internalId != INVALID_ID.toLong()) {
            val intent = LogPlayActivity.createRematchIntent(
                context,
                currentPlay.internalId,
                currentPlay.gameId,
                currentPlay.gameName,
                currentPlay.imageUrl,
                currentPlay.thumbnailUrl,
                currentPlay.heroImageUrl,
                currentPlay.customPlayerSort,
            )
            val pendingIntent = PendingIntent.getActivity(
                context,
                0,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT or if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) PendingIntent.FLAG_IMMUTABLE else 0
            )
            val builder = Action.Builder(
                R.drawable.ic_baseline_replay_24,
                context.getString(R.string.rematch),
                pendingIntent
            )
            return builder.build()
        }
        return null
    }

    companion object {
        const val GEEK_PLAY_URL = "https://boardgamegeek.com/geekplay.php"

        private fun getMapKey(index: Int, key: String) = "players[$index][$key]"
    }
}
